### PR TITLE
Replace TFDV with model monitoring

### PR DESCRIPTION
**Pull request by felix-datatonic**
# Description

Replace TFDV with model monitoring from Vertex AI. 

- [x] removed TFDV components
- [x] add custom batch prediction component
- [x] pass training dataset in BigQuery via shared env. vars.
- [x] update training & prediction pipelines for XGBoost
- [x] update training & prediction pipelines for TensorFlow
- [x] E2E tests  
- [x] improve passing of training dataset between training & prediction pipelines
- [x] update docstrings
- [x] update unit tests

# How has this been tested?

- `make test-all-components`
- `make run PIPELINE_TEMPLATE=xgboost pipeline=training enable_pipeline_caching=false`
- `make run PIPELINE_TEMPLATE=xgboost pipeline=prediction enable_pipeline_caching=false`
- `make run PIPELINE_TEMPLATE=tensorflow pipeline=training enable_pipeline_caching=false`
- `make run PIPELINE_TEMPLATE=tensorflow pipeline=prediction enable_pipeline_caching=false`

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have successfully run the E2E tests, and have included the links to the pipeline runs below
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have updated any relevant documentation to reflect my changes
- [x] I have assigned a reviewer and messaged them

# Pipeline run links:

### CODE REVIEW

File: pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        JobState.JOB_STATE_CANCELLED,
+        JobState.JOB_STATE_EXPIRED,
+    ]
+    job_status_request = GetBatchPredictionJobRequest(name=job_response.name)
+    state = client.get_batch_prediction_job(request=job_status_request).state
+
+    while state not in states:
+        logging.info(f"Batch prediction job state: {state}, waiting for 10s...")
+        time.sleep(10)
+        state = client.get_batch_prediction_job(request=job_status_request).state
```
Comment (browningjp-datatonic) [inline pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py:139]: I think you can just set it running and then return the protobuf as a string - see https://github.com/kubeflow/pipelines/tree/master/components/google-cloud/google_cloud_pipeline_components/proto

File: pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+from pathlib import Path
+from typing import List
+
+
+@component(
+    base_image="python:3.7",
+    packages_to_install=["google-cloud-aiplatform==1.24.1"],
+    output_component_file=str(Path(__file__).with_suffix(".yaml")),
+)
+def export_model(
```
Comment (browningjp-datatonic) [inline pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py:25]: rename

File: pipelines/pipelines/xgboost/training/pipeline.py [hunk 3]
```diff
@@ -48,13 +43,11 @@
...
     project_id: str = os.environ.get("VERTEX_PROJECT_ID"),
     project_location: str = os.environ.get("VERTEX_LOCATION"),
-    pipeline_files_gcs_path: str = os.environ.get("PIPELINE_FILES_GCS_PATH"),
     ingestion_project_id: str = os.environ.get("VERTEX_PROJECT_ID"),
-    tfdv_schema_filename: str = "tfdv_schema_training.pbtxt",
-    tfdv_train_stats_path: str = os.environ.get("TRAIN_STATS_GCS_PATH"),
     model_name: str = "xgboost_with_preprocessing",
     model_label: str = "label_name",
-    dataset_id: str = "preprocessing",
+    dataset_id: str = os.environ.get("BQ_DATASET_NAME"),
```
Comment (browningjp-datatonic) [inline pipelines/pipelines/xgboost/training/pipeline.py:49]: Don't think these should be env variables

File: Makefile [hunk 4]
```diff
@@ -62,18 +62,13 @@ test-all-components: ## Run unit tests for all pipeline components
 		$(MAKE) test-components GROUP=$$(basename $$component_group) ; \
 	done
 
-sync-assets: ## Sync assets folder to GCS. Must specify pipeline=<training|prediction>
```
Comment (browningjp-datatonic) [inline Makefile:65]: We might want to leave the `assets` mechanism in place for other reasons - e.g. a custom dataflow script. WDYT?

**Review by browningjp-datatonic — CHANGES_REQUESTED** [review]

File: pipeline_components/aiplatform/aiplatform/wait_gcp_resources/component.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+@component(
+    base_image="python:3.7",
+    packages_to_install=[
+        "google-cloud-aiplatform>=1.24.0",
+        "google_cloud_pipeline_components>=1.0.0",
+    ],
+    output_component_file=str(Path(__file__).with_suffix(".yaml")),
+)
+def wait_gcp_resources(
```
Comment (browningjp-datatonic) [inline pipeline_components/aiplatform/aiplatform/wait_gcp_resources/component.py:27]: combine into batch prediction component as we discussed

File: pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        "displayName": job_display_name,
+        "model": model.metadata["resourceName"],
+        "inputConfig": input_config,
+        "outputConfig": output_config,
+        "dedicatedResources": {
+            "machineSpec": {"machineType": machine_type},
+            "startingReplicaCount": starting_replica_count,
+            "maxReplicaCount": max_replica_count,
+        },
+        "modelMonitoringConfig": {
```
Comment (browningjp-datatonic) [inline pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py:100]: maybe we just make the whole of the `modelMonitoringConfig` block a dictionary input to the component?

File: pipelines/pipelines/tensorflow/training/assets/tfma_custom_metrics/example_count.py [hunk 7]
```diff
@@ -156,7 +155,6 @@ def merge_accumulators(self, accumulators: Iterable[float]) -> float:
         Args:
             accumulators (Iterable(float)): Accumulator containing the results
                 of every processed chunk
-
```
[the remark this answers is not in the document]
Comment (felix-datatonic) [reply pipelines/pipelines/tensorflow/training/assets/tfma_custom_metrics/example_count.py:159]: pre-commit formatting has applied this, maybe this wasn't done previously?

[the remark this answers is not in the document]
Comment (browningjp-datatonic) [reply pipelines/pipelines/tensorflow/training/assets/tfma_custom_metrics/example_count.py:159]: ok - leave it in if it passes CI linting checks

File: pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py (hunk 6 above)
Comment (felix-datatonic) [reply pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py]: Added three new inputs:
- monitoring_training_dataset: dict = None,
- monitoring_alert_email_addresses: List[str] = None,
- monitoring_skew_config: dict = None,

The first one is passed by `lookup_component` (which won't be used in training in near future due to the CC changes) and the latter two can be set by users in the pipeline.

Comment (browningjp-datatonic) [reply pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py]: Looks good - please can you update the docstring to reflect this

File: pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+from kfp.v2.dsl import Input, Model, component
+from pathlib import Path
+from typing import List, NamedTuple
+
+
+@component(
+    base_image="python:3.7",
+    packages_to_install=[
+        "google-cloud-aiplatform==1.24.1",
+        "google_cloud_pipeline_components==1.0.33",
```
Comment (browningjp-datatonic) [inline pipeline_components/aiplatform/aiplatform/model_batch_predict/component.py:24]: I think this should be `google-cloud-pipeline-components`?

**Review by browningjp-datatonic — CHANGES_REQUESTED** [review]

**Pull request merged**